pkg/inspector/inspector.go

```diff
@@ -0,0 +1,136 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package inspector
+
+import (
+	"context"
+	"sync"
+
+	"github.com/conduitio/conduit/pkg/foundation/log"
+	"github.com/conduitio/conduit/pkg/record"
+	"github.com/google/uuid"
+)
+
+// Session wraps a channel of records and provides:
+// 1. a way to send records to it asynchronously
+// 2. a way to know if it's closed or not
+type Session struct {
+	C chan *record.Record
+
+	logger  log.CtxLogger
+	onClose func()
+	closed  bool
+}
+
+func (s *Session) Close() {
+	s.onClose()
+	close(s.C)
+	s.closed = true
+}
+
+func (s *Session) IsClosed() bool {
+	return s.closed
+}
+
+// send sends a record to the session's channel asynchronously.
+// By sending the record asynchronously we make sure that even
+// if nothing is getting records from the session that we
+// can still continue working.
+func (s *Session) send(r *record.Record) {
+	select {
+	case s.C <- r:
+	default:
+		s.logger.
+			Warn(context.Background()).
+			Msg("session buffer full, record will be dropped")
+
+		return
+	}
+}
+
+// Inspector is attached to an inspectable pipeline component
+// and makes returns records coming in or out of the component.
+// An Inspector is a "proxy" between the pipeline component being
+// inspected and the API, which broadcasts records to all clients.
+type Inspector struct {
+	// sessions is a map of sessions.
+	// keys are sessions IDs.
+	sessions map[string]Session
+	// guards access to sessions
+	lock       sync.Mutex
+	logger     log.CtxLogger
+	bufferSize int
+}
+
+func New(logger log.CtxLogger, bufferSize int) *Inspector {
+	return &Inspector{
+		sessions:   make(map[string]Session),
+		logger:     logger,
+		bufferSize: bufferSize,
+	}
+}
+
+// Send sends the given record to all registered sessions.
+// The method does not wait for consumers to get the records.
+func (i *Inspector) Send(r record.Record) {
+	// copy metadata, to prevent issues when concurrently accessing the metadata
+	var meta record.Metadata
```

Comment:
The thing here is that the OpenCDC version is added (but not always, which is interesting).

Reply:
My theory is that the destination connector calls [`Record.Bytes`](https://github.com/ConduitIO/conduit/blob/e145c81cce73bcb8581d217bd266de225ef2f1a7/pkg/record/record.go#L95) which changes the metadata map in memory (it's a dirty method, a known issue that I think we already discussed when I wrote that code), in the meantime the inspector sends the changed record to the client. This probably only happens in built-in plugins, those operate on the exact same [metadata map](https://github.com/ConduitIO/conduit/blob/e145c81cce73bcb8581d217bd266de225ef2f1a7/pkg/plugin/builtin/v1/internal/toplugin/record.go#L47) as we have internally. Maybe we need to start deep copying maps and payload data even though it will mean more allocations and a bit of a performance hit, but we will keep side effects to a minimum.

Reply:
Oh, this made me think, we have the same issue in fan-out nodes (i.e. if we have multiple destinations) - those also only do a shallow record copy, so changes to metadata and key/payload could be visible in the other destination 😱 

Fanout node clones the message [here](https://github.com/ConduitIO/conduit/blob/e145c81cce73bcb8581d217bd266de225ef2f1a7/pkg/pipeline/stream/fanout.go#L77), which only does a shallow copy of the record [here](https://github.com/ConduitIO/conduit/blob/e145c81cce73bcb8581d217bd266de225ef2f1a7/pkg/pipeline/stream/message.go#L251). Oh we really need a `Record.Clone` method.

Reply:
I've created an issue for this: https://github.com/ConduitIO/conduit/issues/748.